### PR TITLE
Add dma driver for Nios-II MSGDMA core

### PR DESCRIPTION
**Pull request by ramakrishnapallala**

### CODE REVIEW

File: tests/boards/altera_max10/msgdma/src/dma.c [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+	/* Make sure all the data is written out to memory */
+	_nios2_dcache_flush_all();
+
+	/* Start DMA operation */
+	zassert_true(dma_start(dma, chan_id) == 0, "DMA start error");
+
+	while (dma_stat == DMA_OP_STAT_NONE) {
+		k_busy_wait(10);
+	}
```
Comment (rgundi) [inline tests/boards/altera_max10/msgdma/src/dma.c]: I think using a semaphore/event is a more elegant way to achieve this.

File: arch/nios2/core/cache.c [hunk 2]
```diff
@@ -60,3 +60,33 @@ void _nios2_dcache_flush_all(void)
...
+		__asm__ volatile ("initda (%0)" :: "r" (i));
+	}
+
+	/*
+	 * For an unaligned flush request, we've got one more line left.
+	 * Note that this is dependent on ALT_CPU_DCACHE_LINE_SIZE to be a
+	 * multiple of 2 (which it always is).
+	 */
+
+	if (((u32_t) start) & (ALT_CPU_DCACHE_LINE_SIZE - 1)) {
```
Comment (rgundi) [inline arch/nios2/core/cache.c]: I don't understand the purpose of "&"ing start with ALT_CPU_DCACHE_LINE_SIZE - 1. I am not sure i understand what's happening with "unaligned flush request".  May be an example would help?

**Review by rgundi — COMMENTED** [review]
Code looks good but i am seeking a couple of clarifications. Nothing that prevents this from getting approved though.

File: tests/boards/altera_max10/msgdma/src/dma.c (hunk 1 above)
Comment (ramakrishnapallala) [reply tests/boards/altera_max10/msgdma/src/dma.c]: if I use semaphore or mutex the task is put sleep. My intention of the test is to occupy the CPU while DMA is ongoing.

File: arch/nios2/core/cache.c (hunk 2 above)
Comment (ramakrishnapallala) [reply arch/nios2/core/cache.c]: it is checking the DWORD alignment and reused from HAL code.

**Comment by ramakrishnapallala** [comment]
@lpereira @andrewboie Please review this PR.

File: drivers/dma/dma_nios2_msgdma.c [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+		  ALTERA_MSGDMA_DESCRIPTOR_CONTROL_EARLY_TERMINATION_IRQ_MASK;
+
+	/*
+	 * Suppress -Wint-conversion compiler warnings which
+	 * are caused due to the type mismatch b/w the dma_block
+	 * source/destination address (u32_t) and HAL driver's
+	 * descriptor construction API's arguments (alt_32 *)
+	 */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wint-conversion"
```
Comment (andrewboie) [inline drivers/dma/dma_nios2_msgdma.c]: Why do it this way, instead of just casting?

File: drivers/dma/dma_nios2_msgdma.c [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+	if (channel != 0) {
+		SYS_LOG_ERR("Invalid channel number");
+		return -EINVAL;
+	}
+
+	k_sem_take(&cfg->sem_lock, K_FOREVER);
+	status = alt_msgdma_standard_descriptor_async_transfer(cfg->msgdma_dev,
+								&cfg->desc);
+	if (status < 0) {
+		SYS_LOG_ERR("DMA transfer error (%d)", status);
```
Comment (andrewboie) [inline drivers/dma/dma_nios2_msgdma.c]: could move this outside the semaphore take/give section

File: drivers/dma/dma_nios2_msgdma.c [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+	int ret = -EIO;
+	u32_t status;
+
+	k_sem_take(&cfg->sem_lock, K_FOREVER);
+	/* Stop the DMA Dispatcher */
+	IOWR_ALTERA_MSGDMA_CSR_CONTROL(cfg->msgdma_dev->csr_base,
+				       ALTERA_MSGDMA_CSR_STOP_MASK);
+
+	status = IORD_ALTERA_MSGDMA_CSR_STATUS(cfg->msgdma_dev->csr_base);
+	if (status & ALTERA_MSGDMA_CSR_STOP_STATE_MASK) {
```
Comment (andrewboie) [inline drivers/dma/dma_nios2_msgdma.c]: could move this outside the semaphore critical section

File: ext/hal/altera/altera_hal/HAL/src/altera_common.c [hunk 6]
```diff
@@ -80,3 +80,7 @@ unsigned int alt_busy_sleep (unsigned int us)
 	k_busy_wait(us);
 	return 0;
 }
+
+void alt_printf(const char *fmt, ...)
+{
```
Comment (andrewboie) [inline ext/hal/altera/altera_hal/HAL/src/altera_common.c]: what's this function used for? any value in optionally connecting it to printk()?

**Review by tyrantlizard77 — COMMENTED** [review]
Minor stuff

File: ext/hal/altera/altera_hal/HAL/src/altera_common.c (hunk 6 above)
Comment (ramakrishnapallala) [reply ext/hal/altera/altera_hal/HAL/src/altera_common.c]: Just want to provide a wrapper as it is being used few places in the HAL driver.

For now, I would like to keep this as dummy function, we can see if it needed in future to compliment with printk.

File: drivers/dma/dma_nios2_msgdma.c (hunk 4 above)
Comment (ramakrishnapallala) [reply drivers/dma/dma_nios2_msgdma.c]: Ok

File: drivers/dma/dma_nios2_msgdma.c (hunk 5 above)
Comment (ramakrishnapallala) [reply drivers/dma/dma_nios2_msgdma.c]: Ok

File: drivers/dma/dma_nios2_msgdma.c (hunk 3 above)
Comment (ramakrishnapallala) [reply drivers/dma/dma_nios2_msgdma.c]: simple (u32_t *) didn't work, now i have changed it to (alt_u32 *) and it works fine.
I will remove the prama macros. Thanks.

**Comment by ramakrishnapallala** [comment]
@andrewboie, I have addressed the comments and updated the PR. Please have a look.

**Comment by ramakrishnapallala** [comment]
@andyross @andrewboie Can you review the nios2 DMA PR?

**Comment by nashif** [comment]
you need to resolve those conflicts

**Comment by ramakrishnapallala** [comment]
@nashif rebased the PR. Thanks.

**Review by nashif — APPROVED** [review]

**Pull request merged**